### PR TITLE
[SPARK-42630][CONNECT][PYTHON] Introduce UnparsedDataType and delay parsing DDL string until SparkConnectClient is available

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Introduces `UnparsedDataType` and delays parsing DDL string for Python UDFs until `SparkConnectClient` is available.

`UnparsedDataType` carries the DDL string and parse it in the server side.
It should not be enclosed in other data types.

Also changes `createDataFrame` to use the proto `DDLParse`.

### Why are the changes needed?

Currently `parse_data_type` depends on `PySparkSession` that creates a local PySpark, but it won't be available in the client side.

When `SparkConnectClient` is available, we can use the new proto `DDLParse` to parse the data types as string.

### Does this PR introduce _any_ user-facing change?

The UDF's `returnType` attribute could be a string in Spark Connect if it is provided as string.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: python/pyspark/sql/connect/_typing.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ class UserDefinedFunctionLike(Protocol):
     deterministic: bool
 
     @property
-    def returnType(self) -> DataType:
+    def returnType(self) -> DataTypeOrString:
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/_typing.py:60]: do this introduce a behavior change?

https://github.com/apache/spark/blob/6ff760d483124b121d79c3a2d5fdc3ee3f27dd00/python/pyspark/sql/_typing.pyi#L70-L77

It seems that the PySpark UDF's `returnType` is always a `DataType`

cc @xinrong-meng

File: python/pyspark/sql/connect/udf.py [hunk 2]
```diff
@@ -99,9 +97,7 @@ def __init__(
             )
 
         self.func = func
-        self._returnType = (
-            parse_data_type(returnType) if isinstance(returnType, str) else returnType
-        )
+        self._returnType = returnType
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/udf.py:100]: In the vanilla PySpark, a `returnType` property is defined as https://github.com/apache/spark/blob/master/python/pyspark/sql/udf.py#L233, which returns DataType always.

In the existing Connect's code, `wrapper.returnType = self._returnType` is utilized instead.

I am afraid that line of change may break that property parity.

File: python/pyspark/sql/connect/_typing.py (hunk 1 above)
Comment (xinrong-meng) [reply python/pyspark/sql/connect/_typing.py:60]: Oh yes, I commented separately below.

Comment (ueshin) [reply python/pyspark/sql/connect/_typing.py:60]: Yes, that's true.
However, the client is not available here and we don't have a proper way to parse the string here.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/_typing.py:60]: Yeah, but I am still confused about it:

that the old implementation
`PySparkSession.builder.getOrCreate().createDataFrame(data=[], schema=data_type).schema` works.

I also tried 
```
    session = PySparkSession.builder.getOrCreate()
    parsed = session.client._analyze(  # type: ignore[attr-defined]
        method="ddl_parse", ddl_string=data_type
    ).parsed
```
and at least the tests passed.

But if I try
```
     parsed = PySparkSession.builder.getOrCreate().client._analyze(  # type: ignore[attr-defined]
        method="ddl_parse", ddl_string=data_type
    ).parsed
```
the tests always fail with `ValueError: Cannot invoke RPC on closed channel!`

Maybe we will have to add a pure python ddl parser, i don't know

**Review by zhengruifeng — APPROVED** [review]
~~Just one question:~~
~~after creating a udf with ddl string schema, will this ddl string be parsed when calling `udf.returnType`?~~

LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Pull request closed without merging**